src/main/java/com/direwolf20/buildinggadgets/common/tools/BuildingModes.java

```diff
@@ -1,365 +1,279 @@
 package com.direwolf20.buildinggadgets.common.tools;
 
 import com.direwolf20.buildinggadgets.common.config.SyncedConfig;
-import com.direwolf20.buildinggadgets.common.items.gadgets.GadgetBuilding;
 import com.direwolf20.buildinggadgets.common.items.gadgets.GadgetGeneric;
-
-import net.minecraft.block.material.Material;
+import com.direwolf20.buildinggadgets.common.tools.gadget.building.IBuildingMode;
 import net.minecraft.block.state.IBlockState;
 import net.minecraft.entity.player.EntityPlayer;
 import net.minecraft.item.ItemStack;
 import net.minecraft.util.EnumFacing;
+import net.minecraft.util.IStringSerializable;
 import net.minecraft.util.math.AxisAlignedBB;
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 
 import java.util.*;
 
-public class BuildingModes {
-    private static boolean isReplaceable(World world, BlockPos pos, IBlockState setBlock) {
-        if (!setBlock.getBlock().canPlaceBlockAt(world, pos)) {
-            return false;
-        }
-        if (pos.getY() < 0) {
-            return false;
-        }
-        if (SyncedConfig.canOverwriteBlocks) {
-            if (!world.getBlockState(pos).getBlock().isReplaceable(world, pos)) {
-                return false;
-            }
-        } else {
-            if (world.getBlockState(pos).getMaterial() != Material.AIR) {
-                return false;
+public enum BuildingModes implements IStringSerializable {
+
+    BuildToMe((world, player, hit, sideHit, tool) -> {
+        List<BlockPos> coordinates = new ArrayList<>();
+        IBlockState target = GadgetUtils.getToolActualBlock(tool);
+        BlockPos playerPos = new BlockPos(Math.floor(player.posX), Math.floor(player.posY), Math.floor(player.posZ));
+        EnumFacing.Axis axis = sideHit.getAxis();
+        int limit = Math.abs(VectorTools.getAxisValue(playerPos, axis) - VectorTools.getAxisValue(hit, axis));
+
+        // Don'modeImpl place block at where the player stands
+        for (int i = 1; i < limit; i++) {
+            BlockPos pos = hit.offset(sideHit, i);
+            if (isReplaceable(world, pos, target)) {
+                coordinates.add(pos);
             }
         }
-        return true;
-    }
 
-    public static List<BlockPos> getBuildOrders(World world, EntityPlayer player, BlockPos startBlock, EnumFacing sideHit, ItemStack tool) {
-        //GadgetBuilding.ToolMode mode, IBlockState setBlock
-        GadgetBuilding.ToolMode mode = GadgetBuilding.getToolMode(tool);
-        IBlockState setBlock = GadgetUtils.getToolBlock(tool);
+        return coordinates;
+    }),
+    VerticalWall((world, player, hit, sideHit, tool) -> {
```

Comment:
You need to preserve the mode order of the old `GadgetBuilding.ToolMode` enum that this enum replaces. Notice how the names in the radial menu GUI no longer match their icons.

Reply:
if he applys my suggestions, then he could also make them refer to the apropiate textures isntead...

Reply:
No, that would just needlessly change the existing positions of the modes in the radial menu.

Reply:
By applying an id based ordering this could be done easily. Falling back to an enums ordinal/ordering is bad style.

Reply:
I agree that the enum should define it's texture. But it seems:
1. redundant to pass GUI indices in the same order as the value declaration order
2. annoying and counter-intuitive to pass them in an order different from the value declaration order

It makes sense to me to iterate through the values for populating the GUI, and just make them appear in-code in the same order you want them to appear in the GUI. I'd see that as a legitimate use-case for ordering having some small effect. But if it has to be one of the two, then I'd prefer option number 2, which would still mean preserving the order existing value order, in addition to adding GUI indices.

Reply:
I think this is needed, I could quote the whole "Item 31: Use instance fields instead of ordinals" from Effective Java now (The most important point is the maintenance nightmare, which we are noticing now). 
Applying number 2 would be ok, however we'd need to define some kind of other ordering than that, as allowing custom modes will confuse that. However doing a topological sorting as I intended to do it is beyond the scope of this PR, I'll add that afterwards. 

So option 2?

Reply:
Sorry, I meant to type option 1, as that is what the rest of the sentence describes. I've edited it to be correct.

Reply:
ok then option 1...
I have to admit that I had never seen option 2 before, and was kind of curious how it would look...

Reply:
You have marked this as resolved but have not implemented it. At the very least, the mode order of the old `GadgetBuilding.ToolMode` enum that this enum replaces should be preserved.

Reply:
Oh wait I forgot to push XD